### DISCUSSION TITLE
Adding NCEP RAP

### DISCUSSION THREAD

**Discussion by phillipseamore** [open-meteo/open-meteo#188] [General]
Hi,

I noticed that NCEP's hourly Rapid Refresh (RAP) domain [extends into Europe and South-America](https://rapidrefresh.noaa.gov/RAPv4HRRRv3-OpsTrim.jpg):

I'd like to suggest adding that to the GFS processing steps: GFS -> RAP -> HRRR.

**Comment by patrick-zippenfenig** [comment]
Thanks for the tip! Domain seems to be a 13-km with 22 forecast hours. Because of the limited forecast length, I am not sure, if it is adding much benefit for Europe.

I might add it, once my current task list gets a bit shorter ;-)

Data source: https://www.nco.ncep.noaa.gov/pmb/products/rap/

![Screenshot 2023-01-16 at 10 55 32](https://user-images.githubusercontent.com/11670456/212650615-c59abbdb-1d36-457e-ab68-1071507fa170.png)

**Comment by phillipseamore** [comment]
It runs 21 cycles every hour and 51 cycles at 3/9/15/21. Wouldn't this be the only hourly updated model reaching Europe except for METNO?

On a similar note I was looking at the GFS sflux (T1534), global 13km with 384 cycles. It could be interesting doing  a mix of params from GFS sflux and GFS 0.25 > RAP (hourly updated) > HRRR (hourly updated + hires)

**Reply by patrick-zippenfenig** [reply]
51 is definitely more interesting :D I am just not sure if high quality realtime observations from radar and measurement stations in Europe are integrated... It might be worth a try, if the effort to integrate it is not too high

GFS T1534 looks really interesting. Initially I thought it would be some spectral grid encoding that takes ages to decode. With 3072x1536 pixel it is definitely better than the regular 0.25 GFS with 1799x1059 pixel. Do you know if the grid still using a projection? 

![Screenshot 2023-02-09 at 16 37 18](https://user-images.githubusercontent.com/11670456/217859602-0a6d70a1-043b-4efa-93ff-7a0c533529f4.png)

**Reply by phillipseamore** [reply]
According to ncl_filedump the sflux is:
```
       float lat_0 ( lat_0 )
         long_name :    latitude
         grid_type :    Gaussian Latitude/Longitude
         units :        degrees_north
         N :    768
         Di :   0.1171875
         Lo2 :  359.8828
         La2 :  -89.91032
         Lo1 :   0
         La1 :  89.91032
      float lon_0 ( lon_0 )
         long_name :    longitude
         grid_type :    Gaussian Latitude/Longitude
         units :        degrees_east
         N :    768
         Di :   0.1171875
         Lo2 :  359.8828
         La2 :  -89.91032
         Lo1 :   0
         La1 :  89.91032
```

**Reply by phillipseamore** [reply]
Converting to regular lat lon seems to work fine with CDO:
`cdo -sellonlatbox,-180,180,-90,90 -remapbil,r3072x1536 gfs.t18z.sfluxgrbf008.grib2 gfs.t18z.sfluxgrbf008_regular.grib2`

**Reply by patrick-zippenfenig** [reply]
Yeah, I also converted it to NetCDF using CDO copy. In my production code I use the c library bindings for eccodes https://github.com/patrick-zippenfenig/SwiftEccodes. This way I can stream grib files and process grib messages individually with very low memory. It is slightly overkill, but I already encountered 70 GB grib files which worked flawlessly with this approach.

I have a coordinates sanity check that fails right now. It looks like the latitude spacing is mostly `0.117149°`, but at the poles it converges slightly. Because my grid system is defined from the poles onwards, there is an offset in coordinate resolution.



```
-0.11714684008697418
-0.11714661834034246
-0.11714636602391693
-0.11714607724837833
-0.11714574463572092
-0.11714535884183874
-0.11714490789304932
-0.11714437624668506
-0.11714374343191025
-0.11714298204931595
-0.11714205475473705
-0.11714090958722068
-0.1171394725284216
-0.11713763522740805
-0.11713523395285108
-0.11713201178754673
-0.11712754689007454
-0.1171211069251541
-0.1171113280891376
-0.11709542762348235
-0.11706697588064685
-0.11700817943402342
-0.11685317604654699
-0.11616714625951374
```

**Reply by patrick-zippenfenig** [reply]
GFS 0.25 vs T1534. It shows more detail in the alps, but overall very close (as expected). 

![Screenshot 2023-02-09 at 18 36 21](https://user-images.githubusercontent.com/11670456/217893250-aac60122-cfac-4b73-9960-77664ab03e95.png)

I implemented a first downloader in https://github.com/open-meteo/open-meteo/commit/f173697132456126151b6c92456bd088a0ba4da8. Some parameters are not yet supported (precipitation), but I guess for most surface variables it should be drop in replacement for GFS 0.25°. And it has diffuse radiation directly! 

I will wait now for a couple of model runs and see how data looks like ad how much disk space it requires.... Right now my API nodes with 1TB disks are at 92%.....

**Reply by phillipseamore** [reply]
I go out for a few late night beers and you've done it!

Guess the biggest difference might be noticeable along coastlines or large lakes. I feel it's important to try and get the most out of the NOAA data since most of the other models rely too much on ECMWF and I don't like monotony.

We'll need to do something about that diskspace though, especially with all the data that the EU is pushing to get opened and the ensemble models yet to be added. You should put up one of those donation thingies.

**Comment by patrick-zippenfenig** [comment]
GFS T1534 with 13 km is now active by default. GFS025 is still used for upper level variables, but otherwise it is a drop in replacement for the 25 km version.

**Reply by phillipseamore** [reply]
Love it! Thanks.

Regarding the data sources for RAP, I just went over some of the [assimilation scripts](https://github.com/NOAA-EMC/RAP/blob/develop/scripts/) and it's mostly the same [hourly data that is collected for GFS](https://www.nco.ncep.noaa.gov/idsb/nwprod/realtime/index.bufrdump.shtml) plus the latest cycle (for the run hour) from what looks like an internal GFS model run every 3 hours.

**Reply by patrick-zippenfenig** [reply]
Wow... I didn't expect that they still run models with bash scripts that spawn metgrid/wrf/wgrib

**Reply by phillipseamore** [reply]
I was just working with some sflux files again and wanted to double check that we were getting everything we could from them, it looks like sflux is skipped now for APCP and SNOD with 0.25 used instead, even though it's available in sflux.

**Reply by patrick-zippenfenig** [reply]
Ah yes, the small issue was that the running accumulations have varying time windows. Sometimes rain is accumulated of 3 hours, sometimes 12 or longer. Could you create an issue ticket to keep track of it?

Thanks for your detailed validation! It is much appreciated!
